File: go/src/deposit_in_provider/src/deposit_in_provider.go
package deposit_in_provider

import (
	"fmt"

	deposit "interview_mock_deposits_go/deposit/src"
	"interview_mock_deposits_go/deposit/src/domain"
	"interview_mock_deposits_go/internal/events"

	"github.com/sirupsen/logrus"
)

func CreateDepositInProvider(req events.DepositRequestedEvent, requestId string) error {
	depositId := req.Id

	if err := deposit.AcquireIdempotencyLock(depositId); err != nil {
		if err == deposit.ErrLockAlreadyHeld {
			logrus.WithField("depositId", depositId).Info("duplicate event blocked by idempotency lock")
			return nil
		}
		return fmt.Errorf("failed to acquire idempotency lock: %w", err)
	}

	dep, err := deposit.GetDepositById(depositId)
	if err != nil {
		logrus.WithField("error", err.Error()).Warn("error getting deposit")
		return err
	}
	if dep.Status != domain.New {
		logrus.WithFields(logrus.Fields{"depositId": depositId, "status": dep.Status}).Warn("deposit not new")
		return nil
	}
	if dep.UserData == nil {
		logrus.WithField("depositId", depositId).Error("deposit user data is nil")
		return nil
	}

	return PayUser(dep, requestId)
}

File: python/src/deposit_in_provider/deposit_in_provider.py
import logging

from deposit import get_deposit_by_id
from deposit.domain.deposit_status import DepositStatus
from deposit_in_provider.pay_user import pay_user
from internal.events import DepositRequestedEvent

log = logging.getLogger(__name__)


def create_deposit_in_provider(event: DepositRequestedEvent, request_id: str) -> None:
    deposit_id = event.id
    try:
        dep = get_deposit_by_id(deposit_id)
    except Exception as exc:
        log.warning("error getting deposit", extra={"error": str(exc)})
        raise

    if dep.status != DepositStatus.NEW:
        log.warning(
            "deposit not new",
            extra={"depositId": deposit_id, "status": dep.status.name},
        )
        return None

    if dep.user_data is None:
        log.error("deposit user data is nil", extra={"depositId": deposit_id})
        return None

    return pay_user(dep, request_id)
